drafts/2019-10-01-rewriting-irewritable.md

```diff
@@ -0,0 +1,448 @@
+{
+    var children = value.GetChildren();
+    var newChildren = children.Select(transformer);
+    return value.SetChildren(newChildren);
+}
+```
+
+([The real `DefaultRewriteChildren`](https://github.com/benjamin-hodgson/Sawmill/blob/87aea1e5757360b99457c8e2e6a7993fc2176f23/Sawmill/Rewriter.DefaultRewriteChildren.cs) was a bit more complex than this, because it tried to avoid calling `SetChildren` if `transformer` didn't actually change anything.)
+
+In fact, I expect that most `IRewritable`s would just delegate `RewriteChildren` to `DefaultRewriteChildren`. So why did I put `RewriteChildren` on the interface? It's because of Wart 2 --- `GetChildren` can be expensive because it might have to create an `ImmutableList`. (And applying `transformer` to that `ImmutableList` using eg [`ConvertAll`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablelist-1.convertall?view=netcore-3.0) would have to create another one.) So `DefaultRewriteChildren` is slow; the idea of `RewriteChildren` was for objects to transform their children directly, without going via an `ImmutableList`.
+
+```csharp
+class ThreeChildren : IRewritable<ThreeChildren>
+{
+    // ...
+    public ThreeChildren RewriteChildren(Func<ThreeChildren, ThreeChildren> f)
+        => new ThreeChildren(f(_child1), f(_child2), f(_child3));  // no ImmutableList
+}
+```
+
+The extra method makes `IRewritable` harder to understand and harder to implement. Implementing `RewriteChildren` (without using `DefaultRewriteChildren`) can be tricky; if the transformer function doesn't change anything you should avoid recreating the object.
+
+
+## The new way
+
+I'm pleased to say that `IRewritable`'s new design addresses all three of these warts. Remember the "buffer" idea from earlier?
+
+```csharp
+interface IRewritable<T> where T : IRewritable<T>
+{
+    int CountChildren();
+    void GetChildren(Span<T> buffer);
+    T SetChildren(ReadOnlySpan<T> newChildren);
+}
+```
+
+This design is almost identical to the one I outlined earlier --- it uses [`Span`](https://docs.microsoft.com/en-us/dotnet/api/system.span-1?view=netcore-3.0)s instead of arrays. (For the uninitiated, a `Span` is basically a "slice" of an array. It can also be backed by _unmanaged_ memory, though, making it more flexible than `ArraySegment`.)
+
+* `CountChildren` tells Sawmill how much space is needed to copy the children.
+* `GetChildren` copies the current object's immediate children into `buffer`.
+* `SetChildren` creates a copy of the current object with its immediate children replaced.
+
+Here's the important difference. Unlike an array, a `Span` _can't be stored on the heap_. The compiler checks that `Span`s are always confined to the stack. If you pass a `Span` into a method as an argument, you can be confident that the `Span` won't leave the scope of that method's stack frame (just like `ref`).
```

Comment:
Mention that `Span` is a [`ref struct` ](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/ref?view=netcore-3.0#ref-struct-types)and that's why it can't be on the heap?

Maybe something like "Unlike an array, a `Span` is a `ref struct` and _can't be stored on the heap_."